llvm/lib/Analysis/HashRecognize.cpp

```diff
@@ -185,9 +199,14 @@ KnownBits ValueEvolution::computeInstr(const Instruction *I) {
   // Compute the KnownBits for a Select(Cmp()), forcing it to take the branch
   // that is predicated on the (least|most)-significant-bit check.
   CmpPredicate Pred;
-  Value *L, *R, *TV, *FV;
-  if (match(I, m_Select(m_ICmp(Pred, m_Value(L), m_Value(R)), m_Value(TV),
-                        m_Value(FV)))) {
+  Value *L, *R;
+  Instruction *TV, *FV;
+  if (match(I, m_Select(m_ICmp(Pred, m_Value(L), m_Value(R)), m_Instruction(TV),
+                        m_Instruction(FV)))) {
+    Visited.insert(cast<Instruction>(I->getOperand(0)));
+    Visited.insert(TV);
+    Visited.insert(FV);
+
     // We need to check LCR against [0, 2) in the little-endian case, because
     // the RCR check is insufficient: it is simply [0, 1).
     if (!ByteOrderSwapped) {
```

Comment:
The opposite (`if (ByteOrderSwapped)`) path seems to not visit `L`. What if it has side effects?

Reply:
I do not see any side effects (function calls or stores) in the `not.crc.bad.endian.swapped.sb.check` test.

Reply:
Exactly: it's due to `L` being computed instead of `R` (or the reverse), for the `ByteOrderSwapped`.

Reply:
Think about it: what does eliding the significant-bit check effectively do? It means that none of the significant bits are 1, or that we started with a 0 crc (or xor crc data), resulting in a remainder of 0.

Reply:
One of `L` and `R` are instructions, for which we compute KnownBits, and we check the range of `L` and `R` before determining which one we should compute KnownBits for.

Reply:
I believe in this test `L` == `%iv`. You added `call void @print(i16 %crc.next)`. I think we should instead test:
```
%foo = call i32 @foo()
%exit.cond = icmp samesign ult i32 %foo, 7
```
because we don't `compute(L)`.

Reply:
Sorry for the confusion, but I think you're looking at `not.crc.stray.unvisited.call`: the test I'm referring to is `not.crc.stray.unvisited.call` for which `L` is `%xor.crc.data` and `R` is `0`.

Reply:
Oops, indeed in my last comment I looked at a wrong test and a wrong `icmp` (exit cond instead of check sb).

But my point from previous comments remains valid:
```
if (match(I, m_Select(m_ICmp(Pred, m_Value(L), m_Value(R)), m_Instruction(TV),
                        m_Instruction(FV)))) {
```
matches `L`, but there's no `compute(L)`. Can we test and handle a case where L has side effects?